core/src/main/java/org/elasticsearch/action/search/RemoteClusterService.java

```diff
@@ -0,0 +1,395 @@
+                        nodePredicate);
+                    remoteClusters.put(entry.getKey(), remote);
+                }
+
+                // now update the seed nodes no matter if it's new or already existing
+                RemoteClusterConnection finalRemote = remote;
+                remote.updateSeedNodes(entry.getValue(), ActionListener.wrap(
+                    response -> {
+                        if (countDown.countDown()) {
+                            connectionListener.onResponse(response);
+                        }
+                    },
+                    exception -> {
+                        if (countDown.fastForward()) {
+                            connectionListener.onFailure(exception);
+                        }
+                        if (finalRemote.isClosed() == false) {
+                            logger.warn("failed to update seed list for cluster: " + entry.getKey(), exception);
+                        }
+                    }));
+            }
+        }
+        this.remoteClusters = Collections.unmodifiableMap(remoteClusters);
+    }
+
+    /**
+     * Returns <code>true</code> if at least one remote cluster is configured
+     */
+    boolean isCrossClusterSearchEnabled() {
+        return remoteClusters.isEmpty() == false;
+    }
+
+    /**
+     * Groups indices per cluster by splitting remote cluster-alias, index-name pairs on {@link #REMOTE_CLUSTER_INDEX_SEPARATOR}. All
+     * indices per cluster are collected as a list in the returned map keyed by the cluster alias. Local indices are grouped under
+     * {@link #LOCAL_CLUSTER_GROUP_KEY}. The returned map is mutable.
+     *
+     * @param requestIndices the indices in the search request to filter
+     * @param indexExists a predicate that can test if a certain index or alias exists
+     *
+     * @return a map of grouped remote and local indices
+     */
+    Map<String, List<String>> groupClusterIndices(String[] requestIndices, Predicate<String> indexExists) {
+        Map<String, List<String>> perClusterIndices = new HashMap<>();
+        for (String index : requestIndices) {
+            int i = index.indexOf(REMOTE_CLUSTER_INDEX_SEPARATOR);
+            String indexName = index;
+            String clusterName = LOCAL_CLUSTER_GROUP_KEY;
+            if (i >= 0) {
+                String remoteClusterName = index.substring(0, i);
+                if (isRemoteClusterRegistered(remoteClusterName)) {
+                    if (indexExists.test(index)) {
+                        // we use : as a separator for remote clusters. might conflict if there is an index that is actually named
+                        // remote_cluster_alias:index_name - for this case we fail the request. the user can easily change the cluster alias
+                        // if that happens
+                        throw new IllegalArgumentException("Index " + index + " exists but there is also a remote cluster named: "
```

Comment:
I think we need some extra punctuation here, right now this looks like:

`Index foo:bar exists but there is also a remote cluster named: foo can't filter indices`

Perhaps:

`"Can not filter indices; index [foo:bar] exists but there is also a remote cluster named [foo]"`